// react-native-carousel-library/lib/commonjs/components/Pagination.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const {
  width
} = _reactNative.Dimensions.get('screen');

const Pagination = _ref => {
  let {
    scrollRef,
    data,
    paginationStyles: {
      dotSize,
      indicatorColor,
      dotColor,
      top,
      bottom,
      left,
      right
    } = {}
  } = _ref;
  const DOT_SIZE = dotSize ? dotSize : 15;
  const translateX = scrollRef.interpolate({
    inputRange: [-width, 0, width],
    outputRange: [-DOT_SIZE, 0, DOT_SIZE]
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.absoluteWrapper, {
      bottom: bottom ? bottom : 10,
      top: top ? top : undefined,
      left: left ? left : undefined,
      right: right ? right : undefined
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.paginationContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [styles.paginationIndicator, {
      transform: [{
        translateX
      }],
      width: DOT_SIZE,
      height: DOT_SIZE,
      borderRadius: DOT_SIZE / 2,
      borderColor: indicatorColor ? indicatorColor : '#ddd'
    }]
  }), data.map((_, i) => /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.biggerCircle, {
      width: DOT_SIZE
    }],
    key: i
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    key: i,
    style: {
      width: DOT_SIZE * 0.4,
      height: DOT_SIZE * 0.4,
      borderRadius: DOT_SIZE * 0.2,
      backgroundColor: dotColor ? dotColor : '#62a189'
    }
  })))));
};

var _default = Pagination;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  absoluteWrapper: {
    position: 'absolute',
    alignSelf: 'center',
    marginVertical: 20
  },
  paginationContainer: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
    position: 'relative',
    flex: 1
  },
  paginationIndicator: {
    borderWidth: 1,
    position: 'absolute',
    left: 0
  },
  biggerCircle: {
    alignItems: 'center',
    justifyContent: 'center'
  }
});
