// react-native-carousel-library/lib/commonjs/Carousel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Pagination = _interopRequireDefault(require("./components/Pagination"));

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _isURL = _interopRequireDefault(require("validator/lib/isURL"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const {
  width
} = _reactNative.Dimensions.get('screen');

const ITEM_WIDTH = width * 0.76;
const ITEM_HEIGHT = ITEM_WIDTH * 1.47;

const Carousel = _ref => {
  let {
    data,
    landscapeImage,
    pagination = true,
    autoScroll,
    backgroundStyles,
    title,
    imageWrapperStyles,
    titleWrapperStyles,
    titleStyles,
    paginationStyles,
    autoScrollInterval = 5000
  } = _ref;
  const scrollRef = (0, _react.useRef)(new _reactNative.Animated.Value(0)).current;
  const [active, setActive] = (0, _react.useState)(0);
  const viewabilityConfig = {
    itemVisiblePercentThreshold: 50
  };
  const myRef = (0, _react.useRef)(null);
  const onViewableItemsChanged = (0, _react.useCallback)(_ref2 => {
    let {
      viewableItems
    } = _ref2;

    if (viewableItems.length > 0) {
      const {
        index
      } = viewableItems[0];
      setActive(index);
    }
  }, []);
  const viewabilityConfigCallbackPairs = (0, _react.useRef)([{
    viewabilityConfig,
    onViewableItemsChanged
  }]);
  (0, _react.useEffect)(() => {
    let timer;

    if (autoScroll) {
      timer = setInterval(() => {
        var _myRef$current;

        (_myRef$current = myRef.current) === null || _myRef$current === void 0 ? void 0 : _myRef$current.scrollToIndex({
          animated: active === data.length - 1 ? false : true,
          index: active === data.length - 1 ? 0 : active + 1
        });
      }, autoScrollInterval);
    }

    return () => clearInterval(timer);
  }, [active, autoScroll, data.length, autoScrollInterval]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.wrapper, backgroundStyles]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.FlatList, {
    data: data,
    ref: myRef,
    keyExtractor: (_, index) => index.toString(),
    horizontal: true,
    viewabilityConfigCallbackPairs: viewabilityConfigCallbackPairs.current,
    showsHorizontalScrollIndicator: false,
    pagingEnabled: true,
    decelerationRate: "fast",
    initialNumToRender: 2,
    snapToInterval: width,
    onScroll: _reactNative.Animated.event([{
      nativeEvent: {
        contentOffset: {
          x: scrollRef
        }
      }
    }], {
      useNativeDriver: true
    }),
    renderItem: _ref3 => {
      let {
        item,
        index
      } = _ref3;
      const inputRange = [(index - 1) * width, index * width, (index + 1) * width];
      const translateX = scrollRef.interpolate({
        inputRange,
        outputRange: [-width * 0.7, 0, width * 0.7]
      });
      const isUrl = (0, _isURL.default)(item.url + '');
      const imageSource = isUrl ? {
        uri: item.url
      } : item.url;
      return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: [styles.singleItem, imageWrapperStyles]
      }, title && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: [styles.titleContainer, titleWrapperStyles]
      }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
        key: index,
        style: [styles.title, titleStyles]
      }, item.title)), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
        style: [landscapeImage ? styles.horizontalImageContainer : styles.imageContainer]
      }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.Image, {
        source: imageSource,
        style: [landscapeImage ? styles.horizontalImage : styles.image, {
          transform: [{
            translateX
          }]
        }]
      })));
    }
  }), pagination && /*#__PURE__*/_react.default.createElement(_Pagination.default, {
    scrollRef: scrollRef,
    data: data,
    paginationStyles: paginationStyles
  }));
};

var _default = Carousel;
exports.default = _default;

const styles = _reactNative.StyleSheet.create({
  wrapper: {
    width: '100%',
    minHeight: ITEM_WIDTH,
    position: 'relative'
  },
  singleItem: {
    width: width,
    height: ITEM_HEIGHT,
    justifyContent: 'space-around',
    alignItems: 'center'
  },
  imageContainer: {
    width: ITEM_WIDTH,
    height: ITEM_HEIGHT,
    overflow: 'hidden',
    alignItems: 'center',
    borderRadius: 14
  },
  horizontalImageContainer: {
    width: ITEM_HEIGHT - 50,
    height: ITEM_WIDTH,
    overflow: 'hidden',
    alignItems: 'center',
    borderRadius: 14
  },
  image: {
    width: ITEM_WIDTH * 1.2,
    height: ITEM_HEIGHT,
    resizeMode: 'cover'
  },
  horizontalImage: {
    width: ITEM_HEIGHT * 1.2,
    height: ITEM_WIDTH,
    resizeMode: 'cover'
  },
  titleContainer: {
    zIndex: 3,
    position: 'absolute',
    right: 15,
    bottom: -7,
    justifyContent: 'center',
    alignItems: 'center'
  },
  title: {
    color: 'black',
    fontSize: 21,
    paddingBottom: 10,
    letterSpacing: 2
  }
});
